geo/src/algorithm/voronoi_diagram.rs

```diff
@@ -0,0 +1,970 @@
+
+    // Trim the inf_line at the intersection location
+    Some(Line::new(inf_line.start, coord! { x: p_x, y: p_y}))
+}
+
+fn construct_edges_to_inf<T: GeoFloat>(
+    triangles: &[DelaunayTriangle<T>],
+    vertices: &[Coord<T>],
+    edges: &mut Vec<Line<T>>,
+    neighbours: &[(Option<usize>, Option<usize>)],
+    clipping_mask: &Polygon<T>,
+) -> Result<Vec<Line<T>>>
+where
+    f64: From<T>,
+{
+    // Find the mean vertex to determine the direction
+    // of edges going to infinity.
+    let clipping_bounds = clipping_mask
+        .bounding_rect()
+        .ok_or(VoronoiDiagramError::CannotDetermineBoundsFromClipppingMask)?;
+    let clipping_min = clipping_bounds.min();
+    let clipping_max = clipping_bounds.max();
+
+    // Get the vertices with connections to infinity
+    let mut inf_lines: Vec<Line<T>> = Vec::new();
+    for neighbour in neighbours.iter() {
+        if let (None, Some(tri_idx)) = (neighbour.0, neighbour.1) {
+            // Unwrap here is ok as the filtered vec can only be [None, Some(idx)]
+            let inf_edge = define_edge_to_infinity(
+                &triangles[tri_idx],
+                &vertices[tri_idx],
+                edges,
+                &clipping_bounds,
+            )?
+            .ok_or(VoronoiDiagramError::CannotComputeExpectedInfinityVertex)?;
+            let inf_edge_dx_sign = inf_edge.dx().is_sign_positive();
+            let inf_edge_dy_sign = inf_edge.dy().is_sign_positive();
+
+            // Get the clipping mask line where the inf_vertex intersects
+            let intersection_lines: Vec<Line<T>> = clipping_mask
+                .exterior()
+                .lines()
+                .map(|x| trim_line_to_intersection(&inf_edge, &x))
+                .filter(|x| x.is_some())
+                .flatten()
+                // A line can intersect and be outside of the bounding box
+                // filter those out
+                .filter(|x| {
+                    x.end.x >= clipping_min.x
+                        && x.end.x <= clipping_max.x
+                        && x.end.y >= clipping_min.y
+                        && x.end.y <= clipping_max.y
+                })
+                // Get the lines going in the same direction as the inf_edge
+                .filter(|x| {
+                    x.dx().is_sign_positive() == inf_edge_dx_sign
+                        && x.dy().is_sign_positive() == inf_edge_dy_sign
+                })
+                .collect();
+
+            if intersection_lines.len() > 1 {
+                eprintln!("Warning: multiple intersection lines with clipping mask found. Using the first intersection");
+            }
+
+            let intersection_line = intersection_lines
+                .get(0)
+                .ok_or(VoronoiDiagramError::InvalidClippingMaskNoIntersections)?;
+
+            inf_lines.push(*intersection_line);
+        }
+    }
+    Ok(inf_lines)
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum VoronoiDiagramError {
```

Comment:
needs docs for enum and each member

Reply:
I'm going to add the docs as part of the prep for upstream work.